add mkldnn batch_norm backward

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37218 BFloat16: enable prepacked weights's inference
* #37215 BFloat16: add explicit dtype support for to_mkldnn and to_dense
* #37199 add mkldnn relu backward and reshape backward
* **#37147 add mkldnn batch_norm backward**
* #37146 add mkldnn pooling backward
* #36122 add mkldnn linear backward
* #36121 enable mkldnn conv2d backward to support mkldnn tensor input

Differential Revision: [D22440965](https://our.internmc.facebook.com/intern/diff/D22440965)

XiaobingSuper committed:
add mkldnn batch_norm backward

[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/mkldnn/Normalization.cpp:
```diff
@@ -39,35 +53,86 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm(
+  ideep::batch_normalization_backward::compute(
+      x, m, v, grady, w, gradx, gradw, gradb, eps);
+
+  if (weight.is_mkldnn()) {
```
There seems to be a lot of assumptions on the input types here. Can we have the corresponding checks both for the forward and backward functions?

XiaobingSuper replied:
Now, just using second path.

albanD commented on test/test_mkldnn.py:
```diff
@@ -260,20 +260,40 @@ def test_adaptive_avg_pool2d(self):
+                        affine=affine,
+                        track_running_stats=track_running_stats).float().train(train)
+                    if (train or not track_running_stats):
+                        mkldnn_bn = copy.deepcopy(bn)
```
Could you explain why in this case you don't send the module to mkldnn?

XiaobingSuper replied:
For training case, module's parameters awalys a dense tensor, not need to call **mkldnn_utils.to_mkldnn**

albanD replied:
Ho ok. But why is `track_running_stats` here?

albanD commented on test/test_mkldnn.py:
```diff
@@ -260,20 +260,40 @@ def test_adaptive_avg_pool2d(self):
+                        loss1.backward()
+                        loss2.backward()
+                        self.assertEqual(x1.grad, x2.grad.to_dense())
+                        np.testing.assert_allclose(
```
nit: you can use `self.assertEqual(t1, t2, rtol=xx, atol=xx)`.

XiaobingSuper replied:
done.

albanD commented on test/test_mkldnn.py:
```diff
@@ -260,20 +260,40 @@ def test_adaptive_avg_pool2d(self):
+                        loss2.backward()
+                        self.assertEqual(x1.grad, x2.grad.to_dense())
+                        np.testing.assert_allclose(
+                            bn.weight.grad, mkldnn_bn.weight.grad, rtol=1e-3, atol=1e-3)
```
Why is the tolerance so high here give that you assert that y1, y2 are exactly equal below?

XiaobingSuper replied:
diff_weight is computed by `Sum_over_MB*H*W(diff_dst[i] * x_normalized[i])`, the summation order of float number may get different number, MKLDNN will the whole job to a piece job for using muti-thread, the order sum may different with the native path, I also test cuda case, there also has same problem. but for y1, y2, there just has element wise operation, there has't big difference for MKLDNN and native path.

albanD replied:
But given that you work with small Tensors and the only problem is ordering, a tolerance of 1e-5 should be enough no?

albanD commented on test/test_mkldnn.py:
```diff
@@ -260,20 +260,40 @@ def test_adaptive_avg_pool2d(self):
 
-            self._test_serialization(mkldnn_bn, (x.to_mkldnn(),))
-            self._test_tracing(mkldnn_bn, (x.to_mkldnn(),))
+                    self.assertEqual(y1, y2)
```
Could you also check that the running stats have been updated properly?

XiaobingSuper replied:
done.

albanD reviewed (changes requested):
Thanks for the PR!
Just some comments on the test and missing checks.

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"


Differential Revision: [D22440965](https://our.internmc.facebook.com/intern/diff/D22440965)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "add mkldnn batch_norm backward"


Differential Revision: [D22440965](https://our.internmc.facebook.com/intern/diff/D22440965)

[ghstack-poisoned]
